mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt

```diff
@@ -825,42 +868,107 @@ class WidgetAdapter(
+            val rowCount = buttons.maxOfOrNull { it.row ?: 0 } ?: 0
+            val columnCount = min(buttons.maxOfOrNull { it.column ?: 0 } ?: 0, maxColumns)
+
+            // Remove buttons selectively and ensure buttons stay in place when rebinding to the same widget (after
+            // e.g. sending a command on button touch), to make sure touch/release tracking isn't lost in that case
+            buttonViews
+                .filter { (position, buttonView) ->
+                    // Remove buttons beyond the grid size; in case of rebinding to different widgets remove all
+                    // buttons since we *do* want button release tracking to get lost in that case
+                    position.row >= rowCount || position.column >= columnCount ||
+                        (buttonView.tag as? Widget)?.parentId != widget.id
+                }
+                .forEach { (position, buttonView) ->
+                    table.removeView(buttonView)
+                    spareViews.add(buttonView)
+                    buttonViews.remove(position)
+                }
+
+            table.rowCount = rowCount
+            table.columnCount = columnCount
             (0 until table.rowCount).forEach { row ->
                 (0 until table.columnCount).forEach { column ->
-                    val buttonView = spareViews.removeFirstOrNull()
-                        ?: initData.inflater.inflate(R.layout.widgetlist_sectionswitchitem_button, table, false)
-                            as MaterialButton
+                    val buttonView = buttonViews.getOrPut(Position(row, column)) {
+                        val newButton = spareViews.removeFirstOrNull()
+                            ?: initData.inflater.inflate(
+                                R.layout.widgetlist_sectionswitchitem_button,
+                                table,
+                                false
+                            ) as MaterialButton
+
+                        table.addView(
+                            newButton,
+                            GridLayout.LayoutParams(
+                                GridLayout.spec(row, GridLayout.FILL, 1f),
+                                GridLayout.spec(column, GridLayout.FILL, 1f)
+                            )
+                        )
+                        newButton
+                    }
+
                     // Rows and columns start with 1 in Sitemap definition, thus decrement them here
-                    val mapping = mappings.firstOrNull { it.row - 1 == row && it.column - 1 == column }
+                    val button = buttons.firstOrNull { (it.row ?: 0) - 1 == row && (it.column ?: 0) - 1 == column }
                     // Create invisible buttons if there's no mapping so each cell has an equal size
-                    buttonView.isInvisible = mapping == null
-                    if (mapping != null) {
+                    buttonView.isInvisible = button == null
```

Comment:
Shouldn't this also check `button.visibility`, or is this checked elsewhere?

Reply:
This was the original code, I didn't pay too close attention to it

Reply:
Before buttons weren't widgets, this didn't have a visibility flag ;-)

Reply:
the call below (line 941) seems redundant

Reply:
Indeed, agreed.

Reply:
Actually, yes good catch! I'll push a fix.